File: scala/src/main/scala/tennis/game3/TennisGame3.scala
package tennis.game3

import tennis.TennisGame
import tennis.exceptions.UnknownPlayerException

class TennisGame3(val player1Name: String, val player2Name: String) extends TennisGame {

  private var p1Points: Int = 0
  private var p2Points: Int = 0
  private val scoreLabels = Array("Love", "Fifteen", "Thirty", "Forty")

  /**
   * Calculates the score of the game
   * @return the score
   */
  def calculateScore(): String = {
    val isRegular = p1Points < 4 && p2Points < 4 && !(p1Points + p2Points == 6)
    if (isRegular) regularScore(p1Points, p2Points)
    else endGameScore(p1Points, p2Points)
  }

  /**
   * Returns the score for regular situations
   *
   * @param p1Points player 1 points
   * @param p2Points player 2 points
   * @return score string
   */
  private def regularScore(p1Points: Int, p2Points: Int): String = {
    if (p1Points == p2Points) scoreLabels(p1Points) + "-All"
    else scoreLabels(p1Points) + "-" + scoreLabels(p2Points)
  }

  /**
   * Returns the score for end game situations
   *
   * @param p1Points player 1 points
   * @param p2Points player 2 points
   * @return score string
   */
  private def endGameScore(p1Points: Int, p2Points: Int): String = {
    if (p1Points == p2Points) "Deuce"
    else {
      val playerAhead = if (p1Points > p2Points) player1Name else player2Name
      if ((p1Points - p2Points) * (p1Points - p2Points) == 1) "Advantage " + playerAhead else "Win for " + playerAhead
    }
  }

  /**
   * Adds a point to the player score
   * @param playerName player that scored a point
   * @throws [[tennis.exceptions.UnknownPlayerException]]
   */
  @throws(classOf[UnknownPlayerException])
  def wonPoint(playerName: String) {
    if (playerName == "player1")
      p1Points += 1
    else if(playerName == "player2")
      p2Points += 1
    else
      throw UnknownPlayerException(playerName)
  }

}

File: python/src/TennisGame3.py
# -*- coding: utf-8 -*-


class TennisGame3:
    def __init__(self, player1Name, player2Name):
        self.player1Name = player1Name
        self.player2Name = player2Name
        self.pointsPlayer1 = 0
        self.pointsPlayer2 = 0

    def won_point(self, n):
        if n == self.player1Name:
            self.pointsPlayer1 += 1
        else:
            self.pointsPlayer2 += 1

    def score(self):
        if (self.pointsPlayer1 < 4 and self.pointsPlayer2 < 4) and (self.pointsPlayer1 + self.pointsPlayer2 < 6):
            p = ["Love", "Fifteen", "Thirty", "Forty"]
            s = p[self.pointsPlayer1]
            return s + "-All" if (self.pointsPlayer1 == self.pointsPlayer2) else s + "-" + p[self.pointsPlayer2]
        else:
            if self.pointsPlayer1 == self.pointsPlayer2:
                return "Deuce"
            s = self.player1Name if self.pointsPlayer1 > self.pointsPlayer2 else self.player2Name
            return "Advantage " + s if ((self.pointsPlayer1 - self.pointsPlayer2) * (self.pointsPlayer1 - self.pointsPlayer2) == 1) else "Win for " + s
